### PR TITLE
feat: AnyTrust and Celestia support as DA for Arbitrum batches

### PR DESCRIPTION
**Pull request by akolotov**
## Motivation

After this PR, it will be possible to receive AnyTrust Data Availability (DA) information associated with Arbitrum. There is also a first attempt to gather DA information for the case when an Arbitrum rollup uses Celestia as a DA layer, but it was not tested properly due to the absence of such environments.

As part of this PR, the following API endpoints were modified:
- `api/v2/arbitrum/batches/{block_number}`: The `data_availability` section was added with the mandatory field `batch_data_container`, which contains a label of the place where the batch data (blocks and transactions) is available. Possible values are: _null_, `in_blob4844`, `in_calldata`, `in_anytrust`, and `in_celestia`. The rest of the fields in this section depend on the particular DA layer, with its set presented below.
 - `api/v2/block/{block_number}`: The `arbitrum` section was enriched with the batch data container type (`batch_data_container`). This will allow the UI to display the place where the block data is available.
- `api/v2/transactions/{tx_hash}`: The `arbitrum` section was enriched with the batch data container type (`batch_data_container`). This will allow the UI to display the place where the transaction data is available.

Also new API endpoints were added:
- `api/v2/arbitrum/batches/da/anytrust/{data_hash}`: Allows getting batch details by AnyTrust data hash.
- `api/v2/arbitrum/batches/da/celestia/{height}/{tx_commitment }`: Allows getting batch details associated with a Celestia data blob by the transaction commitment and the block height in the Celestia blockchain.

### Sets of the fields in `data_availability` depending on `batch_data_container`

**`in_blob4844` and `in_calldata`**
There are no other fields
 
**`in_anytrust `**
- `data_hash`: The hash of the data in AnyTrust containing the rollup blocks and transactions data.
- `timeout`: The expiration timeout after which data will not be available in AnyTrust.
- `signers`: The list of public keys of committee members who guaranteed data availability, where every pub key record consists of the `key` itself, a `trusted` flag, and the optional `proof` in case trusted is _false_.
- `bls_signature`: The aggregated signature of the committee members.

**`in_celestia `**
- `tx_commitment`: The commitment used in the data blob transaction in the Celestia blockchain.
- `height`: The height of the block in the Celestia blockchain where the data blob transaction was included.

### Enhancements
- Added a new table `arbitrum_da_multi_purpose` to keep data availability records.
- Altered the table `arbitrum_l1_batches` with the new field `batch_container` to specify where batch data is available.
- Modified the behavior of the batches fetcher `Indexer.Fetcher.Arbitrum.TrackingBatchesStatuses` to extract DA info from the `data` field of the `addSequencerL2BatchFromOrigin` call submitting batches to L1.
- Updated `BlockScoutWeb.API.V2.ArbitrumController` and `BlockScoutWeb.API.V2.ArbitrumView` to extend block and transaction views with the new filed, and to extend the batch view with the DA info.

## Checklist for your Pull Request (PR)

  - [ ] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
  - [ ] I checked whether I should update the docs and did so by submitting a PR to https://github.com/blockscout/docs
  - [ ] If I added/changed/removed ENV var, I submitted a PR to https://github.com/blockscout/docs to update the list of env vars at https://github.com/blockscout/docs/blob/master/for-developers/information-and-settings/env-variables.md and I updated the version to `master` in the Version column. Changes will be reflected in this table: https://docs.blockscout.com/for-developers/information-and-settings/env-variables. 
  - [ ] If I added new DB indices, I checked, that they are not redundant with PGHero or other tools.
  - [ ] If I added/removed chain type, I modified the Github CI matrix and PR labels accordingly.

### CODE REVIEW

File: apps/indexer/lib/indexer/fetcher/arbitrum/utils/rpc.ex [hunk 1]
```diff
@@ -114,6 +124,44 @@ defmodule Indexer.Fetcher.Arbitrum.Utils.Rpc do
...
+    ]
+    |> IndexerHelper.read_contracts_with_retries(
+      @selector_sequencer_inbox_contract_abi,
+      json_rpc_named_arguments,
+      @rpc_resend_attempts
+    )
+    |> Kernel.elem(0)
+    |> List.first()
+    |> Kernel.elem(1)
+    |> List.first()
```
Comment (varasev) [inline apps/indexer/lib/indexer/fetcher/arbitrum/utils/rpc.ex:162]: Can you add here a comment which would clarify what tuples are returned so we need to use `Kernel.elem` here?

File: apps/explorer/lib/explorer/chain/arbitrum/reader.ex [hunk 2]
```diff
@@ -910,4 +916,87 @@ defmodule Explorer.Chain.Arbitrum.Reader do
...
+
+    ## Returns
+    - A map containing information about the AnyTrust keyset, otherwise an empty map.
+  """
+  @spec get_anytrust_keyset(binary()) :: map() | nil
+  def get_anytrust_keyset("0x" <> _ = keyset_hash) do
+    get_anytrust_keyset(keyset_hash |> Chain.string_to_block_hash() |> Kernel.elem(1) |> Map.get(:bytes))
+  end
+
+  def get_anytrust_keyset(keyset_hash) do
```
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/explorer/lib/explorer/chain/arbitrum/reader.ex:975]: Thanks, done in a2fd718

File: apps/indexer/lib/indexer/fetcher/zksync/utils/rpc.ex [hunk 3]
```diff
@@ -7,6 +7,8 @@ defmodule Indexer.Fetcher.ZkSync.Utils.Rpc do
 
   alias Indexer.Helper, as: IndexerHelper
 
+  alias Indexer.Helper, as: IndexerHelper
```
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/indexer/lib/indexer/fetcher/zksync/utils/rpc.ex]: Thanks done in 6afa2ae

File: apps/explorer/lib/explorer/chain/arbitrum/batch_block.ex [hunk 4]
```diff
@@ -17,16 +17,31 @@ defmodule Explorer.Chain.Arbitrum.BatchBlock do
...
+                          block is not confirmed yet.
+  """
+  @type to_import :: %{
+          :batch_number => non_neg_integer(),
+          :block_number => non_neg_integer(),
+          :confirmation_id => non_neg_integer() | nil
         }
 
+  @typedoc """
+    * `batch_number` - The number of the Arbitrum batch.
```
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/explorer/lib/explorer/chain/arbitrum/batch_block.ex]: Thanks, done in 1fcbebc

File: apps/explorer/lib/explorer/chain/arbitrum/batch_block.ex [hunk 5]
```diff
@@ -17,16 +17,31 @@ defmodule Explorer.Chain.Arbitrum.BatchBlock do
...
-          batch: %Ecto.Association.NotLoaded{} | L1Batch.t() | nil,
-          block_number: non_neg_integer(),
-          confirmation_id: non_neg_integer() | nil,
-          confirmation_transaction: %Ecto.Association.NotLoaded{} | LifecycleTransaction.t() | nil
+  @typedoc """
+  Descriptor of the a rollup block included in an Arbitrum batch:
+    * `batch_number` - The number of the Arbitrum batch.
+    * `block_number` - The number of the rollup block.
+    * `confirmation_id` - The ID of the confirmation L1 transaction from
+                          `Explorer.Chain.LifecycleTransaction`, or `nil` if the
```
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/explorer/lib/explorer/chain/arbitrum/batch_block.ex]: Addressed after merging with master

File: apps/explorer/lib/explorer/chain/import/runner/arbitrum/da_multi_purpose_records.ex [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    {:ok, inserted}
+  end
+
+  defp default_on_conflict do
+    from(
+      rec in DaMultiPurposeRecord,
+      update: [
+        set: [
+          # don't update `data_key` as it is a primary key and used for the conflict target
+          data_type: fragment("EXCLUDED.data_type"),
```
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/explorer/lib/explorer/chain/import/runner/arbitrum/da_multi_purpose_records.ex:90]: No, with the current approach it is not possible: `data_key` is either the `keyset_hash` or `sha256(celestia_block_height, commitment_tx)`. Due to the cryptographic nature of this data, a collision is almost impossible.

File: apps/indexer/lib/indexer/fetcher/arbitrum/da/anytrust.ex [hunk 7]
```diff
@@ -0,0 +1,414 @@
...
+          %{binary() => boolean()}
+        ) ::
+          {:new_keyset, __MODULE__.keyset(), %{binary() => boolean()}}
+          | {:existing_keyset, nil, %{binary() => boolean()}}
+  defp check_if_new_keyset(keyset_hash, _, cache) when is_map_key(cache, keyset_hash) do
+    {:existing_keyset, nil, cache}
+  end
+
+  defp check_if_new_keyset(keyset_hash, l1_connection_config, cache) do
+    updated_cache = Map.put_new(cache, keyset_hash, true)
```
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/indexer/lib/indexer/fetcher/arbitrum/da/anytrust.ex]: done in fa1e88e

File: apps/indexer/lib/indexer/fetcher/arbitrum/utils/rpc.ex (hunk 1 above)
Comment (akolotov) [reply apps/indexer/lib/indexer/fetcher/arbitrum/utils/rpc.ex:169]: done in 6dadaf6

File: apps/indexer/lib/indexer/fetcher/arbitrum/da/celestia.ex [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    ## Parameters
+    - `source`: The initial list of data to be imported.
+    - `da_info`: The Celestia blob descriptor struct containing details about the data blob.
+
+    ## Returns
+    - An updated list of data structures ready for import, including the Celestia blob descriptor.
+  """
+  @spec prepare_for_import(list(), __MODULE__.t()) :: [Arbitrum.DaMultiPurposeRecord.to_import()]
+  def prepare_for_import(source, %__MODULE__{} = da_info) do
+    key = :crypto.hash(:sha256, :binary.encode_unsigned(da_info.height) <> da_info.tx_commitment)
```
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/indexer/lib/indexer/fetcher/arbitrum/da/celestia.ex]: Done in 7dbb1df

File: apps/explorer/lib/explorer/chain/import/runner/arbitrum/da_multi_purpose_records.ex (hunk 6 above)
[the remark this answers is not in the document]
Comment (varasev) [reply apps/explorer/lib/explorer/chain/import/runner/arbitrum/da_multi_purpose_records.ex:90]: Maybe it makes sense to remove this line then?

**Review by varasev — APPROVED** [review]

File: apps/explorer/priv/arbitrum/migrations/20240527212653_add_da_info.exs [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    create table(:arbitrum_da_multi_purpose, primary_key: false) do
+      add(:data_key, :bytea, null: false, primary_key: true)
+      add(:data_type, :integer, null: false)
+      add(:data, :map, null: false)
+      add(:batch_number, :integer)
+      timestamps(null: false, type: :utc_datetime_usec)
+    end
+
+    create(index(:arbitrum_da_multi_purpose, [:data_type, :data_key]))
+    create(index(:arbitrum_da_multi_purpose, [:data_type, :batch_number]))
```
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/explorer/priv/arbitrum/migrations/20240527212653_add_da_info.exs:23]: As per the [Ecto.Migration documentation](https://hexdocs.pm/ecto_sql/Ecto.Migration.html#index/3) the index name is composed from `table`, `column` as `#{table}_#{column}_index`. So, the names are different. Please confirm.

[the remark this answers is not in the document]
Comment (vbaranov) [reply apps/explorer/priv/arbitrum/migrations/20240527212653_add_da_info.exs:23]: My bad. I misinterpreted the naming.

**Review by vbaranov — APPROVED** [review]

**Pull request merged**